File: frontend/src/api/billing.ts
import { get, post, put } from '@/utils/request'

export type BillingPlanStatus = 'active' | 'archived'
export type BillingInterval = 'month' | 'year' | 'one_time'
export type SubscriptionStatus = 'active' | 'past_due' | 'canceled' | 'expired'
export type BillingOrderStatus = 'pending' | 'paid' | 'canceled' | 'failed'

export interface BillingPlan {
  id: string
  code: string
  name: string
  description: string
  currency: string
  amount_cents: number
  interval: BillingInterval
  entitlements: Record<string, unknown>
  status: BillingPlanStatus
  created_by_user_id: string
  created_at: string
  updated_at: string
}

export interface TenantSubscription {
  id: string
  tenant_id: number
  plan_id: string
  status: SubscriptionStatus
  current_period_start?: string
  current_period_end?: string
  source_order_id?: string
  created_by_user_id: string
  created_at: string
  updated_at: string
  plan?: BillingPlan
}

export interface BillingOrder {
  id: string
  tenant_id: number
  plan_id: string
  provider: string
  provider_order_id: string
  currency: string
  amount_cents: number
  status: BillingOrderStatus
  entitlements: Record<string, unknown>
  created_by_user_id: string
  paid_at?: string
  created_at: string
  updated_at: string
  plan?: BillingPlan
}

export interface TenantBillingStatus {
  tenant_id: number
  plan: BillingPlan
  subscription?: TenantSubscription
  is_free: boolean
}

export interface ApiResponse<T> {
  success: boolean
  data?: T
  message?: string
}

export interface PageResponse<T> extends ApiResponse<T[]> {
  total: number
  page: number
  page_size: number
}

export interface CreateBillingPlanPayload {
  code: string
  name: string
  description?: string
  currency?: string
  amount_cents: number
  interval?: BillingInterval
  entitlements?: Record<string, unknown>
}

export interface UpsertTenantSubscriptionPayload {
  plan_id: string
  status?: SubscriptionStatus
  current_period_start?: string
  current_period_end?: string
  source_order_id?: string
}

export interface CreateBillingOrderPayload {
  tenant_id: number
  plan_id: string
  provider?: string
  provider_order_id?: string
  status?: BillingOrderStatus
  paid_at?: string
}

export function getMyBilling() {
  return get('/api/v1/billing/me') as unknown as Promise<ApiResponse<TenantBillingStatus>>
}

export function listBillingPlans(includeArchived = false) {
  const qs = includeArchived ? '?include_archived=true' : ''
  return get(`/api/v1/system/admin/billing/plans${qs}`) as unknown as Promise<ApiResponse<BillingPlan[]>>
}

export function createBillingPlan(payload: CreateBillingPlanPayload) {
  return post('/api/v1/system/admin/billing/plans', payload) as unknown as Promise<ApiResponse<BillingPlan>>
}

export function listTenantSubscriptions(params: { tenant_id?: number; status?: SubscriptionStatus; page?: number; page_size?: number } = {}) {
  const qs = new URLSearchParams()
  if (params.tenant_id) qs.set('tenant_id', String(params.tenant_id))
  if (params.status) qs.set('status', params.status)
  if (params.page) qs.set('page', String(params.page))
  if (params.page_size) qs.set('page_size', String(params.page_size))
  const suffix = qs.toString()
  return get(`/api/v1/system/admin/billing/subscriptions${suffix ? `?${suffix}` : ''}`) as unknown as Promise<PageResponse<TenantSubscription>>
}

export function upsertTenantSubscription(tenantId: number, payload: UpsertTenantSubscriptionPayload) {
  return put(`/api/v1/system/admin/billing/subscriptions/${tenantId}`, payload) as unknown as Promise<ApiResponse<TenantSubscription>>
}

export function listBillingOrders(params: { tenant_id?: number; status?: BillingOrderStatus; page?: number; page_size?: number } = {}) {
  const qs = new URLSearchParams()
  if (params.tenant_id) qs.set('tenant_id', String(params.tenant_id))
  if (params.status) qs.set('status', params.status)
  if (params.page) qs.set('page', String(params.page))
  if (params.page_size) qs.set('page_size', String(params.page_size))
  const suffix = qs.toString()
  return get(`/api/v1/system/admin/billing/orders${suffix ? `?${suffix}` : ''}`) as unknown as Promise<PageResponse<BillingOrder>>
}

export function createBillingOrder(payload: CreateBillingOrderPayload) {
  return post('/api/v1/system/admin/billing/orders', payload) as unknown as Promise<ApiResponse<BillingOrder>>
}

File: internal/types/billing.go
package types

import "time"

type BillingPlanStatus string

const (
	BillingPlanStatusActive   BillingPlanStatus = "active"
	BillingPlanStatusArchived BillingPlanStatus = "archived"
)

type BillingInterval string

const (
	BillingIntervalMonth   BillingInterval = "month"
	BillingIntervalYear    BillingInterval = "year"
	BillingIntervalOneTime BillingInterval = "one_time"
)

type SubscriptionStatus string

const (
	SubscriptionStatusActive   SubscriptionStatus = "active"
	SubscriptionStatusPastDue  SubscriptionStatus = "past_due"
	SubscriptionStatusCanceled SubscriptionStatus = "canceled"
	SubscriptionStatusExpired  SubscriptionStatus = "expired"
)

type BillingOrderStatus string

const (
	BillingOrderStatusPending  BillingOrderStatus = "pending"
	BillingOrderStatusPaid     BillingOrderStatus = "paid"
	BillingOrderStatusCanceled BillingOrderStatus = "canceled"
	BillingOrderStatusFailed   BillingOrderStatus = "failed"
)

type BillingPlan struct {
	ID              string            `json:"id" gorm:"type:varchar(36);primaryKey"`
	Code            string            `json:"code" gorm:"type:varchar(64);not null;uniqueIndex"`
	Name            string            `json:"name" gorm:"type:varchar(128);not null"`
	Description     string            `json:"description" gorm:"type:text;not null;default:''"`
	Currency        string            `json:"currency" gorm:"type:varchar(16);not null;default:'CNY'"`
	AmountCents     int64             `json:"amount_cents" gorm:"not null;default:0"`
	Interval        BillingInterval   `json:"interval" gorm:"type:varchar(32);not null;default:'month'"`
	Entitlements    JSONMap           `json:"entitlements" gorm:"type:jsonb"`
	Status          BillingPlanStatus `json:"status" gorm:"type:varchar(32);not null;default:'active'"`
	CreatedByUserID string            `json:"created_by_user_id" gorm:"type:varchar(36);not null;default:''"`
	CreatedAt       time.Time         `json:"created_at"`
	UpdatedAt       time.Time         `json:"updated_at"`
}

func (BillingPlan) TableName() string {
	return "billing_plans"
}

type TenantSubscription struct {
	ID                 string             `json:"id" gorm:"type:varchar(36);primaryKey"`
	TenantID           uint64             `json:"tenant_id" gorm:"not null;uniqueIndex"`
	PlanID             string             `json:"plan_id" gorm:"type:varchar(36);not null;index"`
	Status             SubscriptionStatus `json:"status" gorm:"type:varchar(32);not null;default:'active'"`
	CurrentPeriodStart *time.Time         `json:"current_period_start,omitempty"`
	CurrentPeriodEnd   *time.Time         `json:"current_period_end,omitempty"`
	SourceOrderID      *string            `json:"source_order_id,omitempty" gorm:"type:varchar(36);index"`
	CreatedByUserID    string             `json:"created_by_user_id" gorm:"type:varchar(36);not null;default:''"`
	CreatedAt          time.Time          `json:"created_at"`
	UpdatedAt          time.Time          `json:"updated_at"`

	Plan *BillingPlan `json:"plan,omitempty" gorm:"foreignKey:PlanID;references:ID"`
}

func (TenantSubscription) TableName() string {
	return "tenant_subscriptions"
}

type BillingOrder struct {
	ID              string             `json:"id" gorm:"type:varchar(36);primaryKey"`
	TenantID        uint64             `json:"tenant_id" gorm:"not null;index"`
	PlanID          string             `json:"plan_id" gorm:"type:varchar(36);not null;index"`
	Provider        string             `json:"provider" gorm:"type:varchar(32);not null;default:'manual'"`
	ProviderOrderID string             `json:"provider_order_id" gorm:"type:varchar(128);not null;default:'';index"`
	Currency        string             `json:"currency" gorm:"type:varchar(16);not null;default:'CNY'"`
	AmountCents     int64              `json:"amount_cents" gorm:"not null;default:0"`
	Status          BillingOrderStatus `json:"status" gorm:"type:varchar(32);not null;default:'pending'"`
	Entitlements    JSONMap            `json:"entitlements" gorm:"type:jsonb"`
	CreatedByUserID string             `json:"created_by_user_id" gorm:"type:varchar(36);not null;default:''"`
	PaidAt          *time.Time         `json:"paid_at,omitempty"`
	CreatedAt       time.Time          `json:"created_at"`
	UpdatedAt       time.Time          `json:"updated_at"`

	Plan *BillingPlan `json:"plan,omitempty" gorm:"foreignKey:PlanID;references:ID"`
}

func (BillingOrder) TableName() string {
	return "billing_orders"
}

type CreateBillingPlanRequest struct {
	Code         string          `json:"code" binding:"required"`
	Name         string          `json:"name" binding:"required"`
	Description  string          `json:"description"`
	Currency     string          `json:"currency"`
	AmountCents  int64           `json:"amount_cents"`
	Interval     BillingInterval `json:"interval"`
	Entitlements JSONMap         `json:"entitlements"`
}

type UpsertTenantSubscriptionRequest struct {
	PlanID             string             `json:"plan_id" binding:"required"`
	Status             SubscriptionStatus `json:"status"`
	CurrentPeriodStart *time.Time         `json:"current_period_start"`
	CurrentPeriodEnd   *time.Time         `json:"current_period_end"`
	SourceOrderID      *string            `json:"source_order_id"`
}

type CreateBillingOrderRequest struct {
	TenantID        uint64             `json:"tenant_id" binding:"required"`
	PlanID          string             `json:"plan_id" binding:"required"`
	Provider        string             `json:"provider"`
	ProviderOrderID string             `json:"provider_order_id"`
	Status          BillingOrderStatus `json:"status"`
	PaidAt          *time.Time         `json:"paid_at"`
}

type ListBillingOrdersFilter struct {
	TenantID uint64
	Status   BillingOrderStatus
}

type ListTenantSubscriptionsFilter struct {
	TenantID uint64
	Status   SubscriptionStatus
}

type TenantBillingStatus struct {
	TenantID     uint64              `json:"tenant_id"`
	Plan         *BillingPlan        `json:"plan"`
	Subscription *TenantSubscription `json:"subscription,omitempty"`
	IsFree       bool                `json:"is_free"`
}
